Dashboard feature update projects

tremag opened the pull request.

tremag committed:
created custom cell renderer
now i just have to figure out how to dynamically use it

tremag committed:
apparently it must be called "defaultColumn"

tremag committed:
Merge branch 'main' into dashboardFeature_updateProjects

tremag committed:
project dialogue now supports editing

tremag committed:
edit button backend functionality

tremag committed:
edit button frontend functionality

tremag committed:
edit button is now a part of the table

tremag committed:
Merge branch 'main' into dashboardFeature_updateProjects

tremag committed:
fixed a buggy field in the create project popup
i don't think the language was supposed to mirror the link
if i assumed wrong, undo this commit

tremag committed:
added project deletion
now to figure out how to refresh the table :/

tremag committed:
possible fix for table not updating after project creation
as this changes the workflow, a discussion should probably happen with the team/client

tremag committed:
whoops formula was wrong
also clarified some stuff in the todo message for the benefit of future me

KristoferKwan committed:
MERGE main into dashboards branch and added onhover highlighting

KristoferKwan commented on frontend/src/components/Dashboard.js:
```diff
@@ -116,16 +130,50 @@ function Dashboard({history,...props}) {
+      //
+    }
+
+  // custom cell renderer
```
remove comment

KristoferKwan commented on frontend/src/components/Dashboard.js:
```diff
@@ -142,46 +190,78 @@ function Dashboard({history,...props}) {
     return result;
   }
 
   // I don't think this is necessary, since a username will never change without a forced logout and redirect
```
remove comment

KristoferKwan commented on frontend/src/components/DashboardProjectList.js:
```diff
@@ -27,14 +38,22 @@ export default function DashboardProjectList(props) {
                 </TableCell>
               )
             })}
+              {/* <InfoOutlinedIcon
```
is there a reason that delete is inside the modal instead of to the side of the table as an extra column? would say that having the delete directly on the table makes more intuitive sense imo

tremag replied:
ah glad you asked :)
there's a few reasons I decided to do that
1. Your suggestion was actually my first idea, but I initially decided to scrap it because it made the table look a bit crowded. As it is, 1 button on every row is kinda pushing it. If I can figure out how to only make the buttons appear on the row that the user is currently hovering over, then (assuming I can resolve the other issues as well) I may consider reevaluating a separate delete button.
2. As I don't know exactly how pagination works yet and I haven't implemented event delegation yet for the buttons, I'm trying to avoid putting another event listener on every row of the table, as it slows down the page twice as fast.
3. Even though we automate a lot of the client's work, it still takes a lot of time to highlight everything, make classes, and create associations between the different highlights. Because the delete function is irreversible, I figured that burying the option (just slightly) would go a long way towards preventing very unfortunate mistakes. I would like to implement an undo feature (similar to what gmail does right after you send an email), but that's currently near the bottom of my todo list, so I chose to do this as a compromise in the meantime.
4. I'd argue that deletion could be considered a type of modification, which is what that modal popup is for. It's true however that the "i" icon typically means "show more information" as opposed to "modify". I'll have to look into whether a more intuitive alternative exists.

KristoferKwan replied:
Yea thats fine, just thought i'd ask. Ultimately, its up to Rob to decide if he wants any changes, although I imagine he should be fine with the current implementation. As per ur question regarding how you could do it, u could probably hardcode columns (TableCells) into the projectslist in each row and have empty string placeholders in the table header. Again tho, not necessary as I think the general functionality is there. Would say don't put too much emphasis on infinite scrolling -- while it would be nice to have it's not a feature expected by the client, try focusing on making sure all edge cases are considered and also that ur component is well documented.

KristoferKwan commented on frontend/src/components/Dashboard.js:
```diff
@@ -7,16 +7,16 @@ import { useTable } from 'react-table';
 import AddProjectBtn from './layout/AddProjectBtn';
 import './style/Dashboard.css';
 // this function will be an axios call to one of the routes
 // TODO: what is the max number of documents that the database will return?
```
for merge, can u remove all these TODO comments and replace them with comments regarding the functions purpose in finer detail?

KristoferKwan reviewed (changes requested):
added a couple overarching comments

tremag committed:
useEffect was executing an unnecessary number of times
actually should it just wait for openUpdate to change instead?

tremag committed:
made better comments

tremag committed:
updating a project should not redirect the user

tremag committed:
Merge branch 'dashboardFeature_updateProjects' of https://github.com/aSehat/MITR-Project into dashboardFeature_updateProjects

tremag committed:
got rid of (most of) the todo items

tremag committed:
whoops forgot to check the other files
comments should be mostly good now for the dashboard files
also deleted DashboardHeader since it's not being used

KristoferKwan reviewed (approved):
looks good
